### PR TITLE
Resolve Tor control plane address

### LINKED ISSUE #22236: -torcontrol does not resolve a hostname if it is given
**Expected behavior**

-torcontrol should try to resolve a hostname if it is given

**Actual behavior**

The hostname is passed as an ip so the connection fails

```
2021-06-13T20:35:54Z tor: Error parsing socket address mynode.local:9051
2021-06-13T20:35:54Z tor: Initiating connection to Tor control port mynode.local:9051 failed
```

**To reproduce**

Add ```127.0.0.1 mynode``` to /etc/hosts and pass -torcontrol=mynode:9051 as an argument. The connection fails. Try with -torcontrol=127.0.0.1:9051 and it works.

**System information**

Bitcoin Core 0.21.1 from https://bitcoincore.org/bin/bitcoin-core-0.21.1/bitcoin-0.21.1-x86_64-linux-gnu.tar.gz
Ubuntu 18.04

### PR DESCRIPTION
**Pull request by adriansmares**
Closes https://github.com/bitcoin/bitcoin/issues/22236

This PR forces the Tor control plane address to be resolved before a connection attempt is made, similar to how the `-proxy` / `-onion` address is resolved.

The use case for this change is that the control plane may not have a stable address - in a containerized environment perhaps.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/22288/commits/5ed8657cf7c17fb1bc15bfa6ec0f1ab9fe28ec18

Tested this branch on POP!_OS with:

```
$ cat /etc/hosts
127.0.0.1	localghost
```

```
$ cat .bitcoin/bitcoin.conf
torcontrol=localghost:9051
```
Confirmed that it fails on master branch with error:

```
tor: Error parsing socket address localghost:9051
tor: Initializing connection to Tor Control Port localghost:9051 failed
```

**Review by Zero-1729 — APPROVED** [review]
tACK 5ed8657

Tested branch on MacOS 11.3.1.

Used the following configurations in the `/etc/hosts/` and `bitcoin.conf` files respectively:

```sh
$ cat /etc/hosts
127.0.0.1   	zerohost
```

```sh
$ cat bitcoin.conf
torcontrol=zerohost:9051
```

I got the following errors on master (as expected):

```
tor: Error parsing socket address zerohost:9051
tor: Initiating connection to Tor control port zerohost:9051 failed
```

After [this patch](5ed8657), the errors are no longer reported, and the tor connections are established.

Nice first-time contribution and a warm welcome @adriansmares as a contributor! ;)

File: src/torcontrol.cpp [hunk 1]
```diff
@@ -134,12 +134,17 @@ bool TorControlConnection::Connect(const std::string& tor_control_center, const
...
-    // Parse tor_control_center address:port
-    struct sockaddr_storage connect_to_addr;
-    int connect_to_addrlen = sizeof(connect_to_addr);
-    if (evutil_parse_sockaddr_port(tor_control_center.c_str(),
-        (struct sockaddr*)&connect_to_addr, &connect_to_addrlen)<0) {
-        LogPrintf("tor: Error parsing socket address %s\n", tor_control_center);
+
+    CService control_service;
+    if (!Lookup(tor_control_center, control_service, 9051, fNameLookup)) {
+        LogPrintf("tor: Failed to lookup control center %s\n", tor_control_center.c_str());
```
Comment (jonatack) [inline src/torcontrol.cpp:140]: The verb form is "look up".
```suggestion
        LogPrintf("tor: Failed to look up control center %s\n", tor_control_center.c_str());
```

File: src/torcontrol.cpp [hunk 2]
```diff
@@ -134,12 +134,17 @@ bool TorControlConnection::Connect(const std::string& tor_control_center, const
...
+        return false;
+    }
+
+    struct sockaddr_storage control_address;
+    socklen_t control_address_len = sizeof(sockaddr);
+    if (!control_service.GetSockAddr((struct sockaddr*)&control_address, &control_address_len)) {
```
Comment (jonatack) [inline src/torcontrol.cpp:146]: while retouching this, could use a named cast (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast)
```suggestion
    if (!control_service.GetSockAddr(reinterpret_cast<struct sockaddr*>(&control_address), &control_address_len)) {
```

File: src/torcontrol.cpp [hunk 3]
```diff
@@ -153,8 +158,8 @@ bool TorControlConnection::Connect(const std::string& tor_control_center, const
     this->disconnected = _disconnected;
 
     // Finally, connect to tor_control_center
-    if (bufferevent_socket_connect(b_conn, (struct sockaddr*)&connect_to_addr, connect_to_addrlen) < 0) {
-        LogPrintf("tor: Error connecting to address %s\n", tor_control_center);
+    if (bufferevent_socket_connect(b_conn, (struct sockaddr*)&control_address, control_address_len) < 0) {
```
Comment (jonatack) [inline src/torcontrol.cpp:161]: idem
```suggestion
    if (bufferevent_socket_connect(b_conn, reinterpret_cast<struct sockaddr*>(&control_address), control_address_len) < 0) {
```

File: src/torcontrol.cpp [hunk 4]
```diff
@@ -134,12 +134,17 @@ bool TorControlConnection::Connect(const std::string& tor_control_center, const
 {
     if (b_conn)
         Disconnect();
```
Comment (jonatack) [inline src/torcontrol.cpp:136]: Don't hesitate to fix this if you retouch (per doc/developer-notes.md about code style for conditionals)
```diff
-    if (b_conn)
+    if (b_conn) {
         Disconnect();
+    }
```

**Review by jonatack — COMMENTED** [review]
Concept ACK, thanks for looking into this.

File: src/torcontrol.cpp [hunk 5]
```diff
@@ -132,29 +132,36 @@ void TorControlConnection::eventcb(struct bufferevent *bev, short what, void *ct
...
     bufferevent_setcb(b_conn, TorControlConnection::readcb, nullptr, TorControlConnection::eventcb, this);
     bufferevent_enable(b_conn, EV_READ|EV_WRITE);
     this->connected = _connected;
     this->disconnected = _disconnected;
 
     // Finally, connect to tor_control_center
-    if (bufferevent_socket_connect(b_conn, (struct sockaddr*)&connect_to_addr, connect_to_addrlen) < 0) {
-        LogPrintf("tor: Error connecting to address %s\n", tor_control_center);
+    if (bufferevent_socket_connect(b_conn, reinterpret_cast<struct sockaddr*>(&control_address), control_address_len) < 0) {
+        LogPrintf("tor: Error connecting to address %s\n", tor_control_center.c_str());
```
Comment (jonatack) [inline src/torcontrol.cpp:164]: Untested, but ISTM you can omit the 3 uses of `c_str()`

<details><summary>diff</summary><p>

```diff
diff --git a/src/torcontrol.cpp b/src/torcontrol.cpp
index 4a56159228..dcb2175c94 100644
--- a/src/torcontrol.cpp
+++ b/src/torcontrol.cpp
@@ -138,14 +138,14 @@ bool TorControlConnection::Connect(const std::string& tor_control_center, const
 
     CService control_service;
     if (!Lookup(tor_control_center, control_service, 9051, fNameLookup)) {
-        LogPrintf("tor: Failed to look up control center %s\n", tor_control_center.c_str());
+        LogPrintf("tor: Failed to look up control center %s\n", tor_control_center);
         return false;
     }
 
     struct sockaddr_storage control_address;
     socklen_t control_address_len = sizeof(sockaddr);
     if (!control_service.GetSockAddr(reinterpret_cast<struct sockaddr*>(&control_address), &control_address_len)) {
-        LogPrintf("tor: Error parsing socket address %s\n", tor_control_center.c_str());
+        LogPrintf("tor: Error parsing socket address %s\n", tor_control_center);
         return false;
     }
 
@@ -161,7 +161,7 @@ bool TorControlConnection::Connect(const std::string& tor_control_center, const
 
     // Finally, connect to tor_control_center
     if (bufferevent_socket_connect(b_conn, reinterpret_cast<struct sockaddr*>(&control_address), control_address_len) < 0) {
-        LogPrintf("tor: Error connecting to address %s\n", tor_control_center.c_str());
+        LogPrintf("tor: Error connecting to address %s\n", tor_control_center);
         return false;
     }
```
</p></details>

(Optional, you'll save time by squashing the commits after each update, to the final version for review.)

Comment (adriansmares) [reply src/torcontrol.cpp]: Indeed, I see now `LogPrintf` this is based on templates, not untyped varargs.

**Review by jrawsthorne — APPROVED** [review]
tACK 809d02044957b55a2869c0da71ff57656f53ffe4

Thanks for looking into this. This change fixes my original issue

**Comment by prayank23** [comment]
ACK https://github.com/bitcoin/bitcoin/commit/809d02044957b55a2869c0da71ff57656f53ffe4

**Comment by Zero-1729** [comment]
re-ACK 809d02044957b55a2869c0da71ff57656f53ffe4

**Comment by jonatack** [comment]
ACK 809d02044957b55a2869c0da71ff57656f53ffe4 tested various configurations on signet with this branch versus master

File: src/torcontrol.cpp [hunk 6]
```diff
@@ -132,28 +132,35 @@ void TorControlConnection::eventcb(struct bufferevent *bev, short what, void *ct
...
+    }
+
+    CService control_service;
+    if (!Lookup(tor_control_center, control_service, 9051, fNameLookup)) {
+        LogPrintf("tor: Failed to look up control center %s\n", tor_control_center);
+        return false;
+    }
+
+    struct sockaddr_storage control_address;
+    socklen_t control_address_len = sizeof(sockaddr);
```
Comment (theStack) [inline src/torcontrol.cpp:146]:
```suggestion
    struct sockaddr_storage control_address;
    socklen_t control_address_len = sizeof(control_address);
```

**Review by theStack — CHANGES_REQUESTED** [review]
Concept ACK

Welcome as a new contributor adriansmares! 🎉 

Note that on the PR branch connecting to a IPv6 Tor control port doesn't work anymore:
```
$ ./src/bitcoind -torcontrol=::1:9051
...
tor: Error parsing socket address ::1:9051
tor: Initiating connection to Tor control port ::1:9051 failed
...
```

Looking a bit deeper into the problem, it seems like the address length passed to `GetSockAddr()` is too small (16 bytes on my machine), leading to the failure. On all other places in the codebase, `sizeof(sockaddr)` refers to the size of an `sockaddr_storage` instance (always defined one line above), whereas in this case it takes the size of the `sockaddr` structure. See fix below.

**Review by theStack — APPROVED** [review]
ACK cdd51e8ee156f3bb3135be8aa51530a53734153e 🪐 
Tested on Debian bullseye/sid by running `./src/bitcoind -torcontrol=localhost:9051` twice, once with `localhost` set to `127.0.0.1` in `/etc/hosts`, once set to `::1`.

**Comment by jonatack** [comment]
@theStack nice catch.

ACK cdd51e8ee156f3bb3135be8aa51530a53734153e tested various configurations on signet with this branch versus master

**Review by ghost — APPROVED** [review]
> Looking a bit deeper into the problem, it seems like the address length passed to GetSockAddr() is too small (16 bytes on my machine), leading to the failure

Good observation. 

ACK https://github.com/bitcoin/bitcoin/pull/22288/commits/cdd51e8ee156f3bb3135be8aa51530a53734153e

Confirmed it works for ipv6 as well

**Comment by laanwj** [comment]
One argument for not using DNS in the case of Tor is to prevent DNS leaks. I don't think this change, in itself, adds any extra risk of that though.

LGTM ACK cdd51e8ee156f3bb3135be8aa51530a53734153e

**Comment by schildbach** [comment]
I'm currently running into this issue of the Tor control address not being resolved, as I have set up Tor within a separate Docker container. Unfortunately, this PR doesn't seem to have made it into `22.0`. Does someone know a quick workaround to deal with the dynamically changing IP address, until a `22.1` is released?

I could go back to managing the hidden service within Tor (and not use a control port) as I have done in the past, but actually I'd prefer to hand over control to bitcoind.

**Pull request merged**